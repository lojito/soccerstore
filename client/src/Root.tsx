import { type ReactNode } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { createStore } from "redux";
import reducers from "./reducers";

type Props = {
  children: ReactNode;
  initialStoreValue?: {
    favorites: number[];
    cart: number[];
  };
};

// eslint-disable-next-line react-refresh/only-export-components
export default ({ children, initialStoreValue }: Props) => {
  const store = createStore(
    reducers,
    initialStoreValue as unknown as Partial<{ favorites: never; cart: never }>
  );

  return (
    <BrowserRouter>
      <Provider store={store}>{children}</Provider>
    </BrowserRouter>
  );
};
