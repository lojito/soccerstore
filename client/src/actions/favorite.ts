import {
  ADD_TO_FAVORITE,
  FETCH_FAVORITES,
  REMOVE_FROM_FAVORITE,
  REMOVE_ALL_FROM_FAVORITE,
} from "./types";

export type AddToFavoriteAction = {
  type: "add_to_favorite";
  payload: number;
};

export type RemoveFromFavoriteAction = {
  type: "remove_from_favorite";
  payload: number;
};

export type RemoveAllFromFavoriteAction = {
  type: "remove_all_from_favorite";
};

export type FetchFavoritesAction = {
  type: "fetch_favorites";
  payload: number[];
};

export function addFavorite(productId: number): AddToFavoriteAction {
  return {
    type: ADD_TO_FAVORITE,
    payload: productId,
  };
}

export function removeFavorite(productId: number): RemoveFromFavoriteAction {
  return {
    type: REMOVE_FROM_FAVORITE,
    payload: productId,
  };
}

export function removeFavorites(): RemoveAllFromFavoriteAction {
  return {
    type: REMOVE_ALL_FROM_FAVORITE,
  };
}

export function fetchFavorites(productIdList: number[]): FetchFavoritesAction {
  return {
    type: FETCH_FAVORITES,
    payload: productIdList,
  };
}
