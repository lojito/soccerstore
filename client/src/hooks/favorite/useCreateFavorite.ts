import axios from "axios";
import { useCallback, useState } from "react";
import useMounted from "../useMounted";

function useCreateFavorite(token: string) {
  const isMounted = useMounted();
  const [error, setError] = useState("");

  const updateState = useCallback(
    (newError: string) => {
      if (isMounted()) {
        setError(newError);
      }
    },
    [isMounted]
  );

  const handleCreateFavorite = useCallback(
    async (productId: number) => {
      try {
        await axios.post(
          "http://localhost:4000/api/favorites/",
          { productId },
          {
            headers: { Authorization: `Bearer ${token}` },
          }
        );
      } catch (e) {
        updateState("Un error se produjo al crear un nuevo producto favorito.");
      }
    },
    [token, updateState]
  );

  return { error, handleCreateFavorite };
}

export default useCreateFavorite;
