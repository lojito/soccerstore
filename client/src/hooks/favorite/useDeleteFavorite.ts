import axios from "axios";
import { useCallback, useState } from "react";
import useMounted from "../useMounted";

interface DeleteFavorite {
  productId: number;
  error: boolean;
}

function useDeleteFavorite(token: string) {
  const isMounted = useMounted();
  const [state, setState] = useState<DeleteFavorite>({
    productId: 0,
    error: false,
  });

  const updateState = useCallback(
    (newState: DeleteFavorite) => {
      if (isMounted()) {
        setState({ ...newState });
      }
    },
    [isMounted]
  );

  const handleDeleteFavorite = useCallback(
    async (productId: number) => {
      try {
        await axios.delete(`http://localhost:4000/api/favorites/${productId}`, {
          headers: { Authorization: `Bearer ${token}` },
        });
        updateState({
          productId,
          error: false,
        });
      } catch (e) {
        updateState({
          productId,
          error: true,
        });
      }
    },
    [token, updateState]
  );

  return { ...state, handleDeleteFavorite };
}

export default useDeleteFavorite;
