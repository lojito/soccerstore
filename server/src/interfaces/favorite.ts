import { IUser } from "./user";

export interface IFavorite {
  productId: number;
  user: IUser;
}
