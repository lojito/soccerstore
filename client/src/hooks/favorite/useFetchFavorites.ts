import axios from "axios";
import { Dispatch, useCallback, useState } from "react";
import { UnknownAction } from "redux";
import { fetchFavorites } from "../../actions/favorite";
import useMounted from "../useMounted";

interface FetchedFavorites {
  error: boolean;
  isFetching: boolean;
}

function useFetchFavorites(token: string, dispatch: Dispatch<UnknownAction>) {
  const isMounted = useMounted();
  const [state, setState] = useState<FetchedFavorites>({
    error: false,
    isFetching: false,
  });

  const updateState = useCallback(
    (newState: FetchedFavorites) => {
      if (isMounted()) {
        setState({ ...newState });
      }
    },
    [isMounted]
  );

  const handleFetchFavorites = useCallback(async () => {
    try {
      updateState({ error: false, isFetching: true });
      const res = await axios.get("http://localhost:4000/api/favorites", {
        headers: { Authorization: `Bearer ${token}` },
      });
      dispatch(fetchFavorites(res.data));
      updateState({ error: false, isFetching: false });
    } catch (e) {
      updateState({
        error: true,
        isFetching: false,
      });
    }
  }, [token, updateState, dispatch]);

  return { ...state, handleFetchFavorites };
}

export default useFetchFavorites;
