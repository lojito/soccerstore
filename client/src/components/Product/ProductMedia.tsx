import { type FC } from "react";
import { FaHeart } from "react-icons/fa";
import { MdOutlineShoppingCart } from "react-icons/md";
import { useDispatch, useSelector } from "react-redux";
import { styled } from "styled-components";
import { addToCart, removeFromCart } from "../../actions/cart";
import { addFavorite, removeFavorite } from "../../actions/favorite";
import useCreateFavorite from "../../hooks/favorite/useCreateFavorite";
import useDeleteFavorite from "../../hooks/favorite/useDeleteFavorite";
import { RootReducerType } from "../../reducers";

const DivWrapper = styled.div`
  position: relative;
  width: 225px;
  margin: 0 auto;

  & svg.favorite {
    position: absolute;
    right: 15px;
    top: 10px;
    color: red;

    @media screen and (max-width: 1160px) {
      right: 10%;
    }
  }

  & svg.favorite.add {
    color: green;
  }

  & svg.cart {
    position: absolute;
    left: 15px;
    top: 10px;
    color: red;

    @media screen and (max-width: 1160px) {
      left: 10%;
    }
  }

  & svg.cart.added {
    color: green;
  }

  & svg:hover {
    cursor: pointer;
  }

  & image {
    height: 100%;
  }
`;

type Props = {
  id: number;
  image: string;
  isLoggedIn: boolean;
  token: string | null;
};

const ProductMedia: FC<Props> = ({ id, image, isLoggedIn, token }) => {
  const dispatch = useDispatch();

  const productsInFavoritesList = useSelector(
    (state: RootReducerType) => state.favorites
  );
  const isProductInFavoritesList = productsInFavoritesList.includes(id);

  const iconClass = isProductInFavoritesList ? "favorite add" : "favorite";
  const title = isProductInFavoritesList
    ? "Elimina este producto de tu lista de favoritos."
    : "Adiciona este producto a tu lista de favoritos.";

  const { handleCreateFavorite } = useCreateFavorite(token!);
  const { handleDeleteFavorite } = useDeleteFavorite(token!);

  const handleFavorite = async () => {
    if (isProductInFavoritesList) {
      await handleDeleteFavorite(id);
      dispatch(removeFavorite(id));
    } else {
      await handleCreateFavorite(id);
      dispatch(addFavorite(id));
    }
  };

  const productsInCart = useSelector((state: RootReducerType) => state.cart);
  const isProductInCart = productsInCart.includes(id);

  const cartClass = isProductInCart ? "cart added" : "cart";
  const cartTitle = isProductInCart
    ? "Elimina este producto del carrito."
    : "Adiciona este producto al carrito.";

  const handleShoppingCart = async () => {
    if (isProductInCart) {
      dispatch(removeFromCart(id));
    } else {
      dispatch(addToCart(id));
    }
  };

  return (
    <DivWrapper>
      {isLoggedIn && (
        <>
          <FaHeart
            className={iconClass}
            onClick={handleFavorite}
            title={title}
            aria-label="favorite-icon"
            role="img"
          />
          <MdOutlineShoppingCart
            onClick={handleShoppingCart}
            className={cartClass}
            title={cartTitle}
            aria-label="cart-icon"
            role="img"
          />
        </>
      )}
      <img src={image} />
    </DivWrapper>
  );
};

export default ProductMedia;
