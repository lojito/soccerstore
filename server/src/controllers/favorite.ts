import { RequestHandler } from "express";
import { validationResult } from "express-validator";
import Favorite from "../models/favorite";
import User from "../models/user";
import errorHandler from "../middleware/error";
import { ResponseError } from "../interfaces/error";

export const getFavorites: RequestHandler = async (req, res, next) => {
  try {
    const favorites = await Favorite.find(
      { user: req.userId },
      { productId: 1, _id: 0 }
    ).sort({
      createdAt: -1,
    });

    res.status(200).json(favorites.map(({ productId }) => productId));
  } catch (err) {
    errorHandler(err, next);
  }
};

export const createFavorite: RequestHandler = async (req, res, next) => {
  const { productId } = req.body;

  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      const error = new Error("Validation failed.") as ResponseError;
      error.statusCode = 422;
      error.data = errors.array();
      throw error;
    }

    const favorite = await Favorite.create({
      productId,
      user: req.userId,
    });

    const user = await User.findById(req.userId);
    if (user) {
      user.favorites.push(favorite);

      await user.save();

      res.status(201).json({
        message: "Favorite created",
        favorite: {
          productId,
          _id: favorite._id,
        },
      });
    }
  } catch (err: any) {
    errorHandler(err, next);
  }
};

export const deleteFavorite: RequestHandler<{ productId: string }> = async (
  req,
  res,
  next
) => {
  const { productId } = req.params;

  try {
    const favorite = await Favorite.findOne({ productId }, { _id: 1, user: 1 });
    if (!favorite) {
      return res.status(400).json({ error: "No such favorite" });
    }

    if (favorite.user.toString() !== req.userId!.toString()) {
      const error = new Error("Not autorized!") as ResponseError;
      error.statusCode = 403;
      throw error;
    }

    await Favorite.findOneAndRemove({ productId });

    const id = favorite._id;
    const user = await User.findById(req.userId);
    const index = user!.favorites.findIndex(
      (favorite) => favorite.toString() === id.toString()
    );
    if (index !== -1) {
      user!.favorites.splice(index, 1);
      user!.save();
    }

    res.status(200).json(favorite);
  } catch (err: any) {
    errorHandler(err, next);
  }
};
