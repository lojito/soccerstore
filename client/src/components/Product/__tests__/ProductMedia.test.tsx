import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Root from "../../../Root";
import * as useCreateFavorite from "../../../hooks/favorite/useCreateFavorite";
import * as useDeleteFavorite from "../../../hooks/favorite/useDeleteFavorite";
import ProductMedia from "../ProductMedia";

vi.mock("../../../hooks/favorite/useCreateFavorite", () => ({
  default: () => ({
    error: "",
    handleCreateFavorite: vi.fn(),
  }),
}));

vi.mock("../../../hooks/favorite/useDeleteFavorite", () => ({
  default: () => ({
    error: false,
    productId: 0,
    handleDeleteFavorite: vi.fn(),
  }),
}));

describe("ProductMedia", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <Root>
        <ProductMedia
          id={1}
          image="images/balones/1.jpg"
          isLoggedIn={false}
          token=""
        />
      </Root>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should create a new favorite product", async () => {
    const mockHandleCreateFavorite = vi.fn();
    vi.spyOn(useCreateFavorite, "default").mockImplementationOnce(() => ({
      error: "",
      handleCreateFavorite: mockHandleCreateFavorite,
    }));
    const user = userEvent.setup();
    render(
      <Root>
        <ProductMedia id={1} image="images/balones/1.jpg" isLoggedIn token="" />
      </Root>
    );

    await user.click(screen.getByRole("img", { name: "favorite-icon" }));

    expect(mockHandleCreateFavorite).toHaveBeenCalled();
  });

  it("should remove a favorite product", async () => {
    const mockHandleDeleteFavorite = vi.fn();
    vi.spyOn(useDeleteFavorite, "default").mockImplementationOnce(() => ({
      error: false,
      productId: 0,
      handleDeleteFavorite: mockHandleDeleteFavorite,
    }));
    const user = userEvent.setup();
    render(
      <Root initialStoreValue={{ favorites: [1], cart: [] }}>
        <ProductMedia id={1} image="images/balones/1.jpg" isLoggedIn token="" />
      </Root>
    );

    await user.click(screen.getByRole("img", { name: "favorite-icon" }));

    expect(mockHandleDeleteFavorite).toHaveBeenCalled();
  });

  it("should add a new product to the cart", async () => {
    const user = userEvent.setup();
    render(
      <Root>
        <ProductMedia id={1} image="images/balones/1.jpg" isLoggedIn token="" />
      </Root>
    );

    await user.click(screen.getByRole("img", { name: "cart-icon" }));
  });

  it("should remove a product from the cart", async () => {
    const user = userEvent.setup();
    render(
      <Root initialStoreValue={{ favorites: [], cart: [1] }}>
        <ProductMedia id={1} image="images/balones/1.jpg" isLoggedIn token="" />
      </Root>
    );

    await user.click(screen.getByRole("img", { name: "cart-icon" }));
  });
});
