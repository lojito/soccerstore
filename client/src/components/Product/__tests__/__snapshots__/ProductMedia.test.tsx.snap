// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductMedia > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-beySPh imCCJQ"
  >
    <img
      src="images/balones/1.jpg"
    />
  </div>
</DocumentFragment>
`;
